File: leetcode/Medium/javascript/1861. Rotating the Box.js
/**
 * Автор: Дуплей Максим Игоревич - AGLA
 * ORCID: https://orcid.org/0009-0007-7605-539X
 * GitHub: https://github.com/QuadDarv1ne/
 * 
 * Полезные ссылки:
 * 1. Telegram ❃ Хижина программиста Æ: https://t.me/hut_programmer_07
 * 2. Telegram №1 @quadd4rv1n7
 * 3. Telegram №2 @dupley_maxim_1999
 * 4. Rutube канал: https://rutube.ru/channel/4218729/
 * 5. Plvideo канал: https://plvideo.ru/channel/AUPv_p1r5AQJ
 * 6. YouTube канал: https://www.youtube.com/@it-coders
 * 7. ВК группа: https://vk.com/science_geeks
 */

/**
 * Поворачивает матрицу и применяет гравитацию к камням.
 *
 * @param {character[][]} boxGrid - Исходная матрица (боковой вид коробки).
 *        '#' обозначает камень, '*' - препятствие, '.' - пустоту.
 * @return {character[][]} - Матрица после поворота на 90 градусов по часовой
 *         стрелке и применения гравитации.
 */
var rotateTheBox = function(boxGrid) {
    const m = boxGrid.length;
    const n = boxGrid[0].length;

    // Применяем гравитацию в исходной матрице (сдвигаем камни вправо)
    for (let i = 0; i < m; i++) {
        let emptyPos = n - 1;
        for (let j = n - 1; j >= 0; j--) {
            if (boxGrid[i][j] === '*') {
                emptyPos = j - 1;
            } else if (boxGrid[i][j] === '#') {
                // Меняем камень с пустой клеткой
                [boxGrid[i][j], boxGrid[i][emptyPos]] = [boxGrid[i][emptyPos], boxGrid[i][j]];
                emptyPos--;
            }
        }
    }

    // Создаем новую повернутую матрицу (n x m)
    const rotatedBox = Array.from({ length: n }, () => new Array(m));

    // Поворачиваем на 90 градусов по часовой стрелке
    for (let i = 0; i < m; i++) {
        for (let j = 0; j < n; j++) {
            rotatedBox[j][m - 1 - i] = boxGrid[i][j];
        }
    }

    return rotatedBox;
};

File: leetcode/Medium/csharp/1861. Rotating the Box.cs
/**
 * Автор: Дуплей Максим Игоревич - AGLA
 * ORCID: https://orcid.org/0009-0007-7605-539X
 * GitHub: https://github.com/QuadDarv1ne/
 * 
 * Полезные ссылки:
 * 1. Telegram ❃ Хижина программиста Æ: https://t.me/hut_programmer_07
 * 2. Telegram №1 @quadd4rv1n7
 * 3. Telegram №2 @dupley_maxim_1999
 * 4. Rutube канал: https://rutube.ru/channel/4218729/
 * 5. Plvideo канал: https://plvideo.ru/channel/AUPv_p1r5AQJ
 * 6. YouTube канал: https://www.youtube.com/@it-coders
 * 7. ВК группа: https://vk.com/science_geeks
 */

public class Solution {
    /// <summary>
    /// Поворачивает матрицу и применяет гравитацию к камням.
    /// </summary>
    /// <param name="boxGrid">Исходная матрица (вид сбоку).
    /// '#' - камень, '*' - препятствие, '.' - пустота.</param>
    /// <returns>Повёрнутая матрица после падения камней.</returns>
    public char[][] RotateTheBox(char[][] boxGrid) {
        int m = boxGrid.Length;
        int n = boxGrid[0].Length;

        // Этап 1: Симуляция гравитации справа-налево в каждой строке
        for (int i = 0; i < m; ++i) {
            int emptyPos = n - 1;
            for (int j = n - 1; j >= 0; --j) {
                if (boxGrid[i][j] == '*') {
                    emptyPos = j - 1;
                } else if (boxGrid[i][j] == '#') {
                    // Перемещаем камень в доступную позицию emptyPos
                    char temp = boxGrid[i][j];
                    boxGrid[i][j] = boxGrid[i][emptyPos];
                    boxGrid[i][emptyPos] = temp;
                    --emptyPos;
                }
            }
        }

        // Этап 2: Поворот матрицы на 90 градусов по часовой стрелке
        char[][] rotatedBox = new char[n][];
        for (int i = 0; i < n; ++i) {
            rotatedBox[i] = new char[m];
        }

        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                rotatedBox[j][m - 1 - i] = boxGrid[i][j];
            }
        }

        return rotatedBox;
    }
}